### ISSUE TITLE
Function declaration causes completely unrelated code to perform much worse (2x slower)

### ISSUE THREAD

**Issue by Alan-FGR** [dotnet/coreclr#22920]
Hello there. This is my first issue here, I looked for guidelines but I couldn't find anything regarding submitting issues, the closest I could find was the [Contributing Guide](https://github.com/dotnet/coreclr/blob/master/Documentation/project-docs/contributing.md) but it's not about how to file an issue, so I'll just use common sense here :P.

I was benchmarking some code when I suddenly noticed some really strange performance deviations in my results, I managed to pinpoint some of the causes, but what is really going on under the hood is still a mystery to me. In any case I managed to make a simplified repro that you can access [here](https://github.com/Alan-FGR/PrimitiveWrapperBenchmarks/blob/616b30be7a7972f9cc58a839c2a1e619961bcf41/PrimitiveWrapperBenchmarks/Program.cs). I couldn't simplify that code further but I believe it's simple enough. Note that the "Break" preprocessor variable simply guards the declaration of a function that does nothing, but for some weird reason that still causes the benchmark to run two times slower (in release builds) as you can see in the results:

without defining "Break":

| Method |     Mean |     Error |    StdDev |
|------- |---------:|----------:|----------:|
|  First | 28.68 ms | 0.5697 ms | 0.6096 ms |

[results](https://github.com/Alan-FGR/PrimitiveWrapperBenchmarks/tree/616b30be7a7972f9cc58a839c2a1e619961bcf41/PrimitiveWrapperBenchmarks/bin/Release/netcoreapp2.1/results)

with "Break" defined (and no attached debugger just in case although there's no difference):

| Method |     Mean |    Error |   StdDev |
|------- |---------:|---------:|---------:|
|  First | 56.43 ms | 1.081 ms | 1.245 ms |

[results](https://github.com/Alan-FGR/PrimitiveWrapperBenchmarks/tree/616b30be7a7972f9cc58a839c2a1e619961bcf41/PrimitiveWrapperBenchmarks/bin/Release/netcoreapp2.1/results_broken_no_debugger_attached)

I've tried to investigate that issue further, but not only I failed to remove BenchmarkDotNet (BDN) and get the same results (by using a simple StopWatch), but I also couldn't pinpoint what difference in the IL or JIT could justify those results, both the IL and JIT for the "First" (boi I should've rename it) function are exactly the same ([link](https://gist.github.com/Alan-FGR/192c27433210a899e95fa6856f09b72c)) what blew my mind since the loop is inside that function. @saucecontrol suggested it could be alignment, I don't know whether such a difference in performance could happen due to nonalignment, but the code *is* aligned and all `call` instructions in `Main` are calling 8-aligned addresses (but they're calls to BDN).

My first intuition here is that something is wrong with BDN, but I very much doubt it's a bug in it since it's the most popular tool for microbenchmarking and that difference is just too large for something like that to go unnoticed this far. My guess is that there's a bug *somewhere* that BDN somehow triggers and it isn't trivial to reproduce externally. That *somewhere* is the real mystery though, since the native machine code is exactly the same I can only conclude that there's something really hairy going on here, like the code is being invalidated and re-jitted mid-loop or something, but I honestly don't think I'm minimally qualified to even make a reasonable guess here.

Thank you.

**Comment by saucecontrol** [comment]
I tried out your repro but don't see any difference between the two on my system.

```ini
BenchmarkDotNet=v0.11.3, OS=Windows 10.0.17134.556 (1803/April2018Update/Redstone4)
Intel Xeon CPU E3-1505M v6 3.00GHz, 1 CPU, 8 logical and 4 physical cores
Frequency=2929685 Hz, Resolution=341.3336 ns, Timer=TSC
.NET Core SDK=3.0.100-preview-010184
  [Host]     : .NET Core 2.1.8 (CoreCLR 4.6.27317.03, CoreFX 4.6.27317.03), 64bit RyuJIT
  DefaultJob : .NET Core 2.1.8 (CoreCLR 4.6.27317.03, CoreFX 4.6.27317.03), 64bit RyuJIT
```

with #define Break

 Method |     Mean |     Error |    StdDev |
------- |---------:|----------:|----------:|
  First | 35.78 ms | 0.4601 ms | 0.3842 ms |

without

 Method |     Mean |     Error |    StdDev |
------- |---------:|----------:|----------:|
  First | 35.76 ms | 0.7146 ms | 0.8507 ms |

But I see from your results you're running on a much older processor, so code alignment might make more of a difference.  It's not the address of the `call` that would have such an impact but rather the target of frequent jumps, such as the head of your loop

```asm
00007ffb`49192bc0 Repro.First()	
00007ffb`49192bc0 33c0            xor     eax,eax
00007ffb`49192bc2 baffffff00      mov     edx,0FFFFFFh
00007ffb`49192bc7 b901000000      mov     ecx,1
00007ffb`49192bcc 41b801000000    mov     r8d,1
00007ffb`49192bd2 4533c9          xor     r9d,r9d
00007ffb`49192bd5 4103c1          add     eax,r9d
00007ffb`49192bd8 2bc1            sub     eax,ecx
00007ffb`49192bda 3bc1            cmp     eax,ecx
00007ffb`49192bdc 7e02            jle     00007ffb`49192be0
00007ffb`49192bde ffc1            inc     ecx
00007ffb`49192be0 3bc1            cmp     eax,ecx
00007ffb`49192be2 7d03            jge     00007ffb`49192be7
00007ffb`49192be4 41ffc8          dec     r8d
00007ffb`49192be7 3bc1            cmp     eax,ecx
00007ffb`49192be9 7506            jne     00007ffb`49192bf1
00007ffb`49192beb 41b805000000    mov     r8d,5
00007ffb`49192bf1 ffca            dec     edx
00007ffb`49192bf3 413bd1          cmp     edx,r9d
00007ffb`49192bf6 7502            jne     00007ffb`49192bfa
00007ffb`49192bf8 ffc1            inc     ecx
00007ffb`49192bfa 41ffc1          inc     r9d
00007ffb`49192bfd 4181f9ffffff00  cmp     r9d,0FFFFFFh
00007ffb`49192c04 7ccf            jl      00007ffb`49192bd5
00007ffb`49192c06 428d0401        lea     eax,[rcx+r8]
```

In this case, the loop head is at 00007ffb`49192bd5, so definitely not aligned

Related: https://github.com/dotnet/coreclr/issues/11607

**Comment by AndyAyersMS** [comment]
cc @adamsitnik 

This is a case where it would be nice to have access to "CPI" results (cycles per instruction). If the root cause really is loop alignment, we'd expect the fast and slow versions to run the same number of instructions, and the time difference would then just be from internal CPU stalls. 

BDN is already measuring cycles (or a reasonable proxy).

To measure instructions, BDN can make user of [hardware counters](https://benchmarkdotnet.org/articles/samples/IntroHardwareCounters.html) so perhaps we can try this out? If you add in an `InstructionRetired` hardware counter, what does it tell you?

HW counter access might require running as admin, at least on windows. And I think it won't work if you are running in a VM.

**Comment by Alan-FGR** [comment]
@saucecontrol well shouldn't your single core throughput be better though? :confused:
While it's true the loop head isn't aligned, it's not aligned in either version, and in my results you can see that the alignment is identical for all instructions:

Normal | Broken
------------ | -------------
000007fe`793318d0 | 000007fe`795918c0
000007fe`793318d2 | 000007fe`795918c2
000007fe`793318d7 | 000007fe`795918c7
000007fe`793318dc | 000007fe`795918cc
000007fe`793318e2 | 000007fe`795918d2
000007fe`793318e5 | 000007fe`795918d5
000007fe`793318e8 | 000007fe`795918d8
000007fe`793318ea | 000007fe`795918da
000007fe`793318ec | 000007fe`795918dc
000007fe`793318ee | 000007fe`795918de
000007fe`793318f0 | 000007fe`795918e0
000007fe`793318f2 | 000007fe`795918e2
000007fe`793318f4 | 000007fe`795918e4
000007fe`793318f7 | 000007fe`795918e7
000007fe`793318f9 | 000007fe`795918e9
000007fe`793318fb | 000007fe`795918eb
000007fe`79331901 | 000007fe`795918f1
000007fe`79331903 | 000007fe`795918f3
000007fe`79331906 | 000007fe`795918f6
000007fe`79331908 | 000007fe`795918f8
000007fe`7933190a | 000007fe`795918fa
000007fe`7933190d | 000007fe`795918fd
000007fe`79331914 | 000007fe`79591904
000007fe`79331916 | 000007fe`79591906

So the perf difference is definitely not due to alignment at least there... also, 2x slower is a bit too much to be explained simply by misalignment imho.

@AndyAyersMS yes, I'm pretty sure some of the BDN diagnostic outputs should have some better clues on what's happening there. I don't normally use that tool though so I'll have to read some docs in order to turn on more useful diagnostics... when I tried to simply activate some features in the attributes it didn't work so it's not a nobrainer :P... in any case the fact that I simply can't reproduce that issue without BDN makes me suspect there's something fishy going on and maybe that's an issue that wouldn't easily manifest in actual production code, I don't know...

Thank you guys :+1:

**Comment by Alan-FGR** [comment]
Unfortunately it seems "Hardware Counters" aren't gonna happen... I added the `HardwareCounters` attribute and also had to build with the full debug info, first ran it said `BenchmarkDotNet.Diagnostics.Windows` wasn't installed, after installing that though it shows this message:
```
System.ApplicationException: Profile source only availabe on Win8 and beyond.
```
:disappointed:
I'm running as admin but that error is pretty clear

**Comment by Alan-FGR** [comment]
I've just tested on another machine and I could repro the issue, but this time funnily enough defining "break" actually fixes the performance. Here are the results:
[Results with "Break" defined](https://github.com/Alan-FGR/PrimitiveWrapperBenchmarks/blob/18aceaecd3266ba31c3ef6a55c650899be22ae95/haswell_results/results_break_defined/results/Repro-report-github.md)
[Results with "Break" not defined](https://github.com/Alan-FGR/PrimitiveWrapperBenchmarks/blob/18aceaecd3266ba31c3ef6a55c650899be22ae95/haswell_results/results_break_not_defined/results/Repro-report-github.md)
Also, this is something I've mentioned a number of times on gitter, but I think it could be very relevant, in both cases the minimum time of the 'slow' version is *never* less than two times the minimum time of the fast version.
I think that fact could be a decent clue on what's happening, that basically means the processor is doing twice the work in the slow version, if the problem was cache or something else I highly doubt we'll see that same correlation. I ran the tests many times and that rule is never broken. Occasionally there are variations of course, but I think we can assume the minimum time is the fastest the processor can go.
I was suspecting the test was simply running twice, that would explain everything perfectly, but couldn't determine whether that's the case and if it is that sounds like a pretty obvious bug which I highly doubt would go unnoticed.

**Comment by Alan-FGR** [comment]
OK guys, I think I'm getting somewhere with this... I've compiled variations with [one up to 64 declared functions](https://github.com/Alan-FGR/PrimitiveWrapperBenchmarks/blob/tests_without_bdn/PrimitiveWrapperBenchmarks/Program.cs) and managed to repro that with simple stopwatch... if anybody wants to give it a try, I made a [handy script](https://github.com/Alan-FGR/PrimitiveWrapperBenchmarks/blob/tests_without_bdn/build_all_and_run.bat) that should compile and run all the variations so you can see if any of them causes the problem in your machine (atm just compare the numbers manually, should be easy to stop the slow ones... automation coming *someday™*). Later on I'll run some more tests but at least now we can rule out BDN as the source of the problem.

It's really starting to look like a weird alignment issue although I haven't checked the addresses, in this machine the builds that are slow are: 26,27,28,29,30,31,46,47,62,63... there's no clear pattern I can see there.

Boi I just wanted to make some games :cry:

**Comment by Alan-FGR** [comment]
Btw, here's a command for the lazy:
```
git clone https://github.com/Alan-FGR/PrimitiveWrapperBenchmarks.git --single-branch --branch=tests_without_bdn && cd PrimitiveWrapperBenchmarks && build_all_and_run.bat
```
I should've renamed some stuff :P.
If you don't have git installed system wide It should work the same from the bash except you have to add `./` to run the bat file like `./build_all_and_run.bat`. It would be nice of more people could confirm this.

**Comment by adamsitnik** [comment]
> My first intuition here is that something is wrong with BDN, but I very much doubt it's a bug in it since it's the most popular tool for microbenchmarking and that difference is just too large for something like that to go unnoticed this far

thank you for your trust!

> when I tried to simply activate some features in the attributes it didn't work 

I have sent a PR to your repo with a sample config https://github.com/Alan-FGR/PrimitiveWrapperBenchmarks/pull/1/

If you have hit any bugs please repor an issue in https://github.com/dotnet/BenchmarkDotNet

>  I'm running as admin but that error is pretty clea

This is a limitation we get from TraceEvent library, which gets it from Windows OS ;(

**Comment by adamsitnik** [comment]
I have run the benchmark with following config:

```cs
public static void Main(string[] args) =>
    BenchmarkRunner.Run<Repro>(
        DefaultConfig.Instance
            .With(Job.Default
                .WithLaunchCount(1) // how many processes to run
                .WithId("NO BREAK"))
            .With(Job.Default
                .WithLaunchCount(1) // how many processes to run
                .WithCustomBuildConfiguration("BREAK")
                .With(new [] { new MsBuildArgument("/p:Optimize=true") }) // we are using custom build configuration, so we need to tell MSBuild in explicit way to optimize the code
                .WithId("BREAK"))
            // .With(HardwareCounter.InstructionRetired)
            // .With(new EtwProfiler(new EtwProfilerConfig()))
            .With(DisassemblyDiagnoser.Create(DisassemblyDiagnoserConfig.Asm)));
```

and **surprisingly I was able to reproduce the problem only once** on my PC (with the first run):

```ini
BenchmarkDotNet=v0.11.4, OS=Windows 10.0.17763.253 (1809/October2018Update/Redstone5)
Intel Xeon CPU E5-1650 v4 3.60GHz, 1 CPU, 12 logical and 6 physical cores
.NET Core SDK=3.0.100-preview-009812
  [Host]   : .NET Core 2.1.6 (CoreCLR 4.6.27019.06, CoreFX 4.6.27019.05), 64bit RyuJIT
  BREAK    : .NET Core 2.1.6 (CoreCLR 4.6.27019.06, CoreFX 4.6.27019.05), 64bit RyuJIT
  NO BREAK : .NET Core 2.1.6 (CoreCLR 4.6.27019.06, CoreFX 4.6.27019.05), 64bit RyuJIT
```

| Method |      Job |        Arguments | BuildConfiguration |     Mean |     Error |    StdDev |
|------- |--------- |----------------- |------------------- |---------:|----------:|----------:|
|  First |    BREAK | /p:Optimize=true |              BREAK | 35.78 ms | 0.2611 ms | 0.2442 ms |
|  First | NO BREAK |          Default |            Default | 17.86 ms | 0.0965 ms | 0.0903 ms |

The Disassembly Diagnoser reported following ASM (**in both cases the body of the loop was 16 byte aligned**):

```asm
00007ff8`da8131c0 Repro.First()
00007ff8`da8131c0 33c0            xor     eax,eax
00007ff8`da8131c2 baffffff00      mov     edx,0FFFFFFh
00007ff8`da8131c7 b901000000      mov     ecx,1
00007ff8`da8131cc 41b801000000    mov     r8d,1
00007ff8`da8131d2 4533c9          xor     r9d,r9d
00007ff8`da8131d5 4103c1          add     eax,r9d
00007ff8`da8131d8 2bc1            sub     eax,ecx
00007ff8`da8131da 3bc1            cmp     eax,ecx
00007ff8`da8131dc 7e02            jle     00007ff8`da8131e0
00007ff8`da8131de ffc1            inc     ecx
00007ff8`da8131e0 3bc1            cmp     eax,ecx
00007ff8`da8131e2 7d03            jge     00007ff8`da8131e7
00007ff8`da8131e4 41ffc8          dec     r8d
00007ff8`da8131e7 3bc1            cmp     eax,ecx
00007ff8`da8131e9 7506            jne     00007ff8`da8131f1
00007ff8`da8131eb 41b805000000    mov     r8d,5
00007ff8`da8131f1 ffca            dec     edx
00007ff8`da8131f3 413bd1          cmp     edx,r9d
00007ff8`da8131f6 7502            jne     00007ff8`da8131fa
00007ff8`da8131f8 ffc1            inc     ecx
00007ff8`da8131fa 41ffc1          inc     r9d
00007ff8`da8131fd 4181f9ffffff00  cmp     r9d,0FFFFFFh
00007ff8`da813204 7ccf            jl      00007ff8`da8131d5
00007ff8`da813206 428d0401        lea     eax,[rcx+r8]
```

```asm

00007ff8`da832c70 Repro.First()
00007ff8`da832c70 33c0            xor     eax,eax
00007ff8`da832c72 baffffff00      mov     edx,0FFFFFFh
00007ff8`da832c77 b901000000      mov     ecx,1
00007ff8`da832c7c 41b801000000    mov     r8d,1
00007ff8`da832c82 4533c9          xor     r9d,r9d
00007ff8`da832c85 4103c1          add     eax,r9d
00007ff8`da832c88 2bc1            sub     eax,ecx
00007ff8`da832c8a 3bc1            cmp     eax,ecx
00007ff8`da832c8c 7e02            jle     00007ff8`da832c90
00007ff8`da832c8e ffc1            inc     ecx
00007ff8`da832c90 3bc1            cmp     eax,ecx
00007ff8`da832c92 7d03            jge     00007ff8`da832c97
00007ff8`da832c94 41ffc8          dec     r8d
00007ff8`da832c97 3bc1            cmp     eax,ecx
00007ff8`da832c99 7506            jne     00007ff8`da832ca1
00007ff8`da832c9b 41b805000000    mov     r8d,5
00007ff8`da832ca1 ffca            dec     edx
00007ff8`da832ca3 413bd1          cmp     edx,r9d
00007ff8`da832ca6 7502            jne     00007ff8`da832caa
00007ff8`da832ca8 ffc1            inc     ecx
00007ff8`da832caa 41ffc1          inc     r9d
00007ff8`da832cad 4181f9ffffff00  cmp     r9d,0FFFFFFh
00007ff8`da832cb4 7ccf            jl      00007ff8`da832c85
00007ff8`da832cb6 428d0401        lea     eax,[rcx+r8]
```

**Comment by adamsitnik** [comment]
@AndyAyersMS you are right, the number of instructions retired is almost the same with every run:

| Method |      Job |     Mean | InstructionRetired/Op |
|------- |--------- |---------:|----------------------:|
|  First |    BREAK | 17.56 ms |           252,295,833 |
|  First | NO BREAK | 17.25 ms |           252,291,667 |

**Comment by adamsitnik** [comment]
the most suprising thing is that I was able to reproduce it once, even with `WithLaunchCount(10)` I was not able to repro it

@AndreyAkinshin do you have any ideas what could have caused this?

**Comment by Alan-FGR** [comment]
@adamsitnik wow, thank you for the input! :)
I'm relieved somebody else could repro this tbh... it's really strange though that you can't reproduce it consistently, here it happens every single time.
As I said I could repro the *"phenomena"* :smirk: without BDN so it's definitely not the case for a bug report about it. Also, the minor difficulty I was having was because it needs the full symbols to print out the code along the asm and I passed the `true` that activates that by mistake... the tool is solid and very easy to use, you guys did a great job!
I'll take a look at that PR @adamsitnik, thank you.

I think the new branch that doesn't run BDN could be more helpful to debug that though... that script will build separate binaries for each function declaration count, and since the code is simpler it should be easier to figure what exactly is happening differently... I just don't know what kind of tools would be appropriate for that... maybe some tools that allows you to not only dump the whole jitted code but maybe also step through and whatnot... and perhaps similar tools for IL code too.

**Comment by AndyAyersMS** [comment]
The loop has some data-dependent branches in it, so I wonder if we're seeing inadvertent collisions in the HW branch predictor. @adamsitnik if you can repro, try looking at how branch misprediction rates differ in slow and fast runs.

**Comment by Alan-FGR** [comment]
OK so I just ran the tests on the other machine and the slow ones are the exact same (26,27,28,29,30,31,46,47,62,63) so at least there's consistency... Also, I understand that some of you might not think this is a big deal and this could be dismissed as some hardware peculiarity but I'd like to at least know what's going on here.
I would like to step through the machine code but that wasn't possible using x64dbg because of the way code is invoked... one idea I have is to compile using CoreRT, if the same issue happens then I think I'd be able to use the debugging tools I'm familiar with to figure out what's going on. Since f25 is good and f26 is bad I'm using them for comparison (so it's easy to spot the differences in a diffviewer).
Thank you guys.

**Comment by Alan-FGR** [comment]
Hey guys, so I could reproduce the issue in CoreRT and after some scripts and debugging it really seems @saucecontrol was right, I've compiled 32 versions and so far the observations have consistently shown that all the 'slow' versions have addresses ending in `4`... the strange part though is that addresses ending in `C` which aren't 8-aligned are fast.
I'm using x64dbg and I'm checking the addresses manually what's very boring :P.
I still don't understand why the BDN version was showing aligned addresses though :confused:.
In any case maybe this could provide some solid ground to at very least request a force-alignment feature... most of the popular C++ compilers have a feature like that and I believe GCC in max optimization does that by default. I thought instructions would be at least 8-aligned by default.

**Comment by Alan-FGR** [comment]
I forgot a very important note: since the problem is alignment one would assume that if you have a lot of functions then some would be aligned and some don't, but that's not what happens... I've tried adding 64 functions to the same binary and they're always all the same, that is they're either all slow or all fast, so the problem is something is happening that's breaking alignment apparently for everything (need more tests though)

**Comment by adamsitnik** [comment]
@Alan-FGR I think that the best thing to do here would be to use Intel VTune (it is free now) and run the produced exe with it.

To produce an .exe that Intel VTune can run you should take your project (the one that is not using BenchmarkDotNet), use `--dotnet publish -c Release` and then tell VTune to run the produced exe

**Comment by Alan-FGR** [comment]
WHAT? VTune is free? That's awesome! :smile:
As @saucecontrol said on gitter, the head alignment per se doesn't matter, the fact 4 is slow and C is fast is probably because it's offsetting the 'problem' instruction into its sweet spot. I'll definitely run VTune in order to try and find it.

**Comment by Alan-FGR** [comment]
Oh boi... actually you need to be working in a company to download VTune free and I'm currently not :cry:

**Comment by Alan-FGR** [comment]
NOTE: I've been chatting a lot with @saucecontrol on gitter so I apologize if some of this sounds a bit out of context.

@saucecontrol I don't know why BDN shows aligned addresses when the problem manifests, is it guaranteed that the addresses are the real thing? Are those addresses collected or are they calculated? I could try to compile the BDN code with CoreRT to check if they're correct but I guess that won't be easy since it's a complex piece of code...

I've double checked the addresses and the binaries and yes I'm 100% positive they are 64-bit (not only process explorer (which is my "Task Manager") says they're 64-bit images but the plat spec `50 45 00 00 64 86` is there in the header). For the addresses I've made [this gist](https://gist.github.com/Alan-FGR/8d553adf397b6f910c8ad854d097f223) that contains x64dbg output, first column is memory address.
I ran VTune and the 'broken' cases are front-end bound with 100% L2 replacement and twice the clocks per instruction.

This whole thing makes me question the usefulness of benchmarking tbh... or at least using time as a measure for benchmarks... from what I've seen here if you can't align the addresses to produce consistency then the time could be completely irrelevant, maybe a better metric for benchmarking would be clocks and instructions and maybe you could even have an estimation of the instruction costs on different hardware so you have a better idea of what variations to expect in real world due to hardware peculiarities.

**Comment by Alan-FGR** [comment]
btw, here's the VTune output if you guys want to take a look: https://gist.github.com/Alan-FGR/757558f9d481654ae618d7778cc119e4

**Comment by AndyAyersMS** [comment]
Does this VTune output come from running on Core? If so, I'm puzzled. We shouldn't see those kind of method alignments.

**Comment by AndyAyersMS** [comment]
Or maybe you're running on x86 (and not x64)?

**Comment by Alan-FGR** [comment]
@AndyAyersMS nope, it's definitely x64, I didn't even build x86 versions. And those addresses are bigger than 32-bit can handle unless I'm mistaken (0x140005164 for example has 9 hex digits).
Not sure what you meant by "Core" there (I'm n00b :P )... it's running on the i5. In any case I'm attaching not only the VTune projects (which include the result `.amplxe`) but also the binaries in case you guys want to see for yourselves:
[CoreRT Binaries.zip](https://github.com/dotnet/coreclr/files/2923454/CoreRT.Binaries.zip)
[VTune Projects and Results.zip](https://github.com/dotnet/coreclr/files/2923453/VTune.Projects.and.Results.zip)
Let me know if anything else might be helpful.

**Comment by AndyAyersMS** [comment]
> Not sure what you meant by "Core" there 

Sorry, I meant, were you measuring via VTune while running on CoreCLR?

But it looks like the answer is no -- you were running on CoreRT. I wonder if CoreRT is doing anything to ensure method entry alignment. Would guess not? cc @jkotas 

Your original repro was on CoreCLR where we'd expect on x64 to always get the same 16 byte entry alignment. And since method body code doesn't vary between fast & slow, then we have same relative branch offsets. So perhaps we're seeing some 32 bit alignment related artifact?

For the CoreRT data -- I recall reading somewhere that on Intel CPUs, code having 3 conditional branches in a 16 byte aligned block could also cause perf issues, but I can't figure out where I read that; it is not there in recent Intel opt guides. @fiigii ring any bells? Note in the "slow" version you have exactly this; in the "fast" version the first in-loop branch is in a different 16 byte block.

**Comment by Alan-FGR** [comment]
@AndyAyersMS oh I think you mean Core as in CoreCLR/RT as opposed to .Net Framework? If so, yes those results are from the CoreRT binaries, also note that x64dbg results are consistent with VTune as well as the actual measurements.

**Comment by Alan-FGR** [comment]
@AndyAyersMS OK I was ninjaed there :laughing:... I'll run CoreCLR on VTune to see if I get those alignments.

**Comment by Alan-FGR** [comment]
OK, nothing makes sense anymore... here are the results from CoreCLR:
https://gist.github.com/Alan-FGR/52254794fdbcd75b51cd898ecf27c516
The numbers are comparable to the CoreRT ones (twice the clocks and all) but the alignment is the same :(... just when stuff was starting to make sense :P

Here are the VTune projects and results:
[VTune Projects and Results CoreCLR.zip](https://github.com/dotnet/coreclr/files/2923511/VTune.Projects.and.Results.CoreCLR.zip)

Also, note that while on CoreRT the problem manifests on define `f0` (so I'm using f0 (bad) and f1 (good) for comparison), on CoreCLR the first define is `f26` so I'm comparing f25 (good) to f26 (bad). This is mostly to make it easier to run diffviewers and whatnot.

**Comment by Alan-FGR** [comment]
Btw, just so you guys know I'm getting perfect consistency here, this is with the binaries acquired and built with `git clone https://github.com/Alan-FGR/PrimitiveWrapperBenchmarks.git --single-branch --branch=tests_without_bdn && cd PrimitiveWrapperBenchmarks && build_all_and_run.bat` as described [here](https://github.com/dotnet/coreclr/issues/22920#issuecomment-468707764):
```PowerShell
PS C:\Projects\_temp\PrimitiveWrapperBenchmarks\tests> for($i = 0; $i -lt 20; $i++) {iex ("dotnet "+(25+$i%2)+"\PrimitiveWrapperBenchmarks.dll")}
Running f25
Took 430 ms
Running f26
Took 820 ms
Running f25
Took 431 ms
Running f26
Took 792 ms
Running f25
Took 460 ms
Running f26
Took 787 ms
Running f25
Took 440 ms
Running f26
Took 810 ms
Running f25
Took 452 ms
Running f26
Took 815 ms
Running f25
Took 434 ms
Running f26
Took 821 ms
Running f25
Took 428 ms
Running f26
Took 804 ms
Running f25
Took 452 ms
Running f26
Took 829 ms
Running f25
Took 435 ms
Running f26
Took 838 ms
Running f25
Took 444 ms
Running f26
Took 836 ms
```
And the binaries built with CoreRT:
```PowerShell
PS C:\Projects\_temp\PrimitiveWrapperBenchmarks\PrimitiveWrapperBenchmarks> for($i = 0; $i -lt 20; $i++) {iex ("f"+($i%2)+"\PrimitiveWrapperBenchmarks.exe")}
Running f0
Took 835 ms
Running f1
Took 452 ms
Running f0
Took 831 ms
Running f1
Took 424 ms
Running f0
Took 861 ms
Running f1
Took 423 ms
Running f0
Took 842 ms
Running f1
Took 425 ms
Running f0
Took 847 ms
Running f1
Took 417 ms
Running f0
Took 818 ms
Running f1
Took 426 ms
Running f0
Took 849 ms
Running f1
Took 429 ms
Running f0
Took 824 ms
Running f1
Took 420 ms
Running f0
Took 856 ms
Running f1
Took 431 ms
Running f0
Took 856 ms
Running f1
Took 426 ms
```

**Comment by Alan-FGR** [comment]
Btw, I made [this spreadsheet](https://docs.google.com/spreadsheets/d/18m5sdaRLvgYARcXt01t3dd316r_Jo0-i3bhPQ-4s-Yw/edit?usp=sharing) with the CoreCLR results because that gist is confusing... green is the good/fast version.

**Comment by Alan-FGR** [comment]
OK guys, so it seems the problem is still alignment. I've made a table with all the CoreCLR addresses and you can clearly see a correlation there (B is the 'broken'/slow version):
```Text
0  - 0x7fe6f6e2df0
1  - 0x7fe6f6e2df0
2  - 0x7fe6f642df0
3  - 0x7fe6f6b2df0
4  - 0x7fe6f642df0
5  - 0x7fe6f6d2df0
6  - 0x7fe6f632df0
7  - 0x7fe6f6d2df0
8  - 0x7fe6f662df0
9  - 0x7fe6f6a2df0
10 - 0x7fe6f662df0
11 - 0x7fe6f6a2df0
12 - 0x7fe6f652df0
13 - 0x7fe6f6d2df0
14 - 0x7fe6f642df0
15 - 0x7fe6f6a2df0
16 - 0x7fe6f632df0
17 - 0x7fe6f6b2df0
18 - 0x7fe6f642df0
19 - 0x7fe6f6b2df0
20 - 0x7fe6f862df0
21 - 0x7fe6f6f2df0
22 - 0x7fe6f102df0
23 - 0x7fe6f6c2df0
24 - 0x7fe6f122df0
25 - 0x7fe6f6d2df0
26 B 0x7fe6f102e00
27 B 0x7fe6f6f2e00
28 B 0x7fe6f102e00
29 B 0x7fe6f6d2e00
30 B 0x7fe6f0f2e00
31 B 0x7fe6f6e2e00
32 - 0x7fe6ed63210
33 - 0x7fe6ddd3210
34 - 0x7fe6ed630d0
35 - 0x7fe6ddc30d0
36 - 0x7fe6ed730d0
37 - 0x7fe6ddd30d0
38 - 0x7fe6f6d30d0
39 - 0x7fe6f0f30d0
40 - 0x7fe6f6f30d0
41 - 0x7fe6f1130d0
42 - 0x7fe6f6e30d0
43 - 0x7fe6f1030d0
44 - 0x7fe6f6f30d0
45 - 0x7fe6f1130d0
46 B 0x7fe6f6e30c0
47 B 0x7fe6f1130c0
48 - 0x7fe6f6e30d0
49 - 0x7fe6f1230d0
50 - 0x7fe6f6d2ef0
51 - 0x7fe6f0f2ef0
52 - 0x7fe6f6c2ef0
53 - 0x7fe6f112ef0
54 - 0x7fe6f6f2ef0
55 - 0x7fe6f0f2ef0
56 - 0x7fe6f6e2ef0
57 - 0x7fe6f102ef0
58 - 0x7fe6f6f2ef0
59 - 0x7fe6f102ef0
60 - 0x7fe6f6d32f0
61 - 0x7fe6f1232f0
62 B 0x7fe6f6f3020
63 B 0x7fe6f103020
```
So it seems @saucecontrol was right again, the problem is instructions being pushed out of their sweet spot in terms of alignment, I just couldn't imagine that it would happen beyond 16 bytes but apparently it does.

**Comment by saucecontrol** [comment]
That was actually what I noticed about the asm dumps from the time @adamsitnik was able to repro it.  Good case started at `0x...70` and bad started at `0x...C0`

That makes sense if you consider the layout of the instruction and μop caches in the processor.  The μop cache is 32 bytes, and you want the idea portion of the method to fit in that cache for branch prediction and such to work best.

Imagine your method is 64 bytes of code, and it's aligned to 16 bytes.  Half the time (statistically) it will look like |32|32| becuase it's aligned to 32 bytes as well.  The other half, it looks like |16|32|16|.  If the code that needs to be in the μop cache happens to be the middle bit of the method, the second layout would be good and the first layout bad.  That's basically what you're seeing.

**Comment by Alan-FGR** [comment]
@saucecontrol I think that is exactly what's happening! To be honest I had no idea that 32 aligning was a thing... I always trusted my compilers to be smart enough so I never really cared about instruction alignment except when *really* necessary like SIMD.
In any case it's a pity we talked so much on gitter but that information won't be easily accessible by other people in the future... something there might help somebody else facing a similar situation.

I still would like to have more control over alignment in CoreCLR but especially in CoreRT which for some reason aligns to 4 what always breaks everything.
In any case, everything makes some sense again... you can see in the spreadsheet how the hot loops end up in different 32 alignment when the top is unaligned. Albeit not perfectly aligned, now I know exactly what's causing the performance drop, so at least from a software development perspective it's all perfectly explainable now, and specific architecture peculiarities are definitely off-topic here :P.

I can't thank you guys enough for helping me on this and in special @saucecontrol with whom I learned a lot of stuff and who's definitely a badass :smile:.

So do you guys think this issue would provide some grounds for an alignment feature request?

**Comment by jkotas** [comment]
> CoreRT which for some reason aligns to 4 what always breaks everything.

The most compact code runs the fastest on average. The large code alignment helps the hottest methods, but hurts everything else. This is especially true for ahead-of-time compilation where it is hard to guess methods that are important (without profile guide optimization).

It is trivial to change the alignment in CoreRT if you would like to experiment: https://github.com/dotnet/corert/blob/c7df360bcf89465739dc9af98633e3d420152ea6/src/JitInterface/src/CorInfoImpl.cs#L249

**Comment by Alan-FGR** [comment]
@jkotas well I'd surely like to experiment with that number. Why did you link that line and not [this one](https://github.com/dotnet/corert/blob/3c31103c264e93d036e1c15833edc37f4a83be74/src/Common/src/TypeSystem/Common/TargetDetails.cs#L173) though? Should I just pass a value in that line you linked instead of changing the actual TargetDetails.MinimumFunctionAlignment? Also, any plans to soft code that? Maybe read from some config file or parameter?

**Comment by MichalStrehovsky** [comment]
> Why did you link that line and not [this one](https://github.com/dotnet/corert/blob/3c31103c264e93d036e1c15833edc37f4a83be74/src/Common/src/TypeSystem/Common/TargetDetails.cs#L173) though?

That would also change the alignment of some assembly stubs. Those are tiny and giving them a generous alignment would probably hurt (code loses locality for no reason).

FWIW, for .NET Native on UWP, we default to 16-byte alignment based on some measurements on microbenchmarks that the UTC backend team did in the past (the C++ compiler also defaults to this number).

We ended up having to offer an opt out that switches alignment to 4 for customers that have tight size budgets (like inbox Windows UWP apps).

I think we should just do the same thing for the CoreRT compiler - default to 16 and have an opt out command line option in ILC that makes it use the minimum allowed (pull requests welcome).

**Comment by AndyAyersMS** [comment]
I believe we can make a case for selectively increasing alignment demand to 32 bytes for some methods (say based on tiering feedback, or perhaps attributes). This is not the first time we've seen things like this (#16873).

Having the jit optimize alignment of internal branch targets (#11607) is a likely prerequisite. And internal alignment is tricky to do for 16 bytes, so will be even trickier to do for 32.

Am going to label this as future as I don't see us making any CoreCLR changes here for 3.0.
